StreamEventJournalPTest.smokeTest: snapshots not fully restored

## Error Message

processor output in mode "snapshots enabled, restoring every snapshot" doesn't match expected:<0
1
2
3> but was:<0
2
1>


java.lang.AssertionError
	at com.hazelcast.jet.core.test.JetAssert.throwNotEqual(JetAssert.java:67)
	at com.hazelcast.jet.core.test.JetAssert.assertEquals(JetAssert.java:50)
	at com.hazelcast.jet.core.test.TestSupport.runTest(TestSupport.java:582)
	at com.hazelcast.jet.core.test.TestSupport.expectOutputs(TestSupport.java:334)
	at com.hazelcast.jet.core.test.TestSupport.expectOutput(TestSupport.java:302)
	at com.hazelcast.jet.impl.connector.StreamEventJournalPTest.smokeTest(StreamEventJournalPTest.java:107)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:98)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)

## Standard Output

Finished Running Test: when_lostItems_afterRestore in 15.151 seconds.
13:51:45,171  INFO || - [JobExecutionService] hz._hzInstance_6_jet.cached.thread-5 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Execution plan for jobId=1d75-6ebe-5e2d-19f1, jobName='1d75-6ebe-5e2d-19f1', executionId=1c1b-edaf-840f-df43 initialized
13:51:45,188 DEBUG || - [MasterContext] hz._hzInstance_6_jet.async.thread-3 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Init of job '1d75-6ebe-5e2d-19f1', execution 1c1b-edaf-840f-df43 was successful
13:51:45,188 DEBUG || - [MasterContext] hz._hzInstance_6_jet.async.thread-3 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Executing job '1d75-6ebe-5e2d-19f1', execution 1c1b-edaf-840f-df43
13:51:45,191  INFO || - [JobExecutionService] hz._hzInstance_6_jet.async.thread-3 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Start execution of job '1d75-6ebe-5e2d-19f1', execution 1c1b-edaf-840f-df43 from coordinator [127.0.0.1]:5701
13:51:45,222 DEBUG || - [Job] when_processorsWithNoPartitions_then_snapshotRestoreWorks - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Sending RESTART_GRACEFUL request for job 1d75-6ebe-5e2d-19f1 (name ??)
13:51:45,224  INFO || - [MasterContext] when_processorsWithNoPartitions_then_snapshotRestoreWorks - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Starting snapshot 0 for job '1d75-6ebe-5e2d-19f1', execution 1c1b-edaf-840f-df43, terminal
13:51:45,227 DEBUG || - [JobCoordinationService] hz._hzInstance_6_jet.async.thread-3 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] job '1d75-6ebe-5e2d-19f1', execution 1c1b-edaf-840f-df43 snapshot is scheduled in 200000ms
13:51:45,250  INFO || - [LifecycleService] Thread-0 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
13:51:45,251  INFO || - [MigrationManager] hz._hzInstance_4_jet.cached.thread-2 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Shutdown request of Member [127.0.0.1]:5701 - 25bd1fc9-6c32-40a6-b6ac-3627474b35ef this is handled
13:51:45,252  INFO || - [Node] Thread-0 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Shutting down connection manager...
13:51:45,252  INFO || - [Node] Thread-0 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Shutting down node engine...
13:51:45,444 DEBUG || - [AsyncSnapshotWriterImpl] hz._hzInstance_6_jet.jet.cooperative.thread-0 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Stats for src: keys=2, chunks=1, bytes=89
13:51:45,446 DEBUG || - [SnapshotOperation] hz._hzInstance_6_jet.jet.cooperative.thread-0 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Snapshot 0 for job '1d75-6ebe-5e2d-19f1', execution 1c1b-edaf-840f-df43 finished successfully on member
13:51:45,509 DEBUG || - [JobExecutionService] hz._hzInstance_6_jet.jet.cooperative.thread-0 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Execution of job '1d75-6ebe-5e2d-19f1', execution 1c1b-edaf-840f-df43 completed with failure
java.util.concurrent.CompletionException: com.hazelcast.jet.impl.exception.TerminatedWithSnapshotException
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:604)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at com.hazelcast.jet.impl.util.NonCompletableFuture.internalComplete(NonCompletableFuture.java:55)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$ExecutionTracker.taskletDone(TaskletExecutionService.java:396)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.dismissTasklet(TaskletExecutionService.java:332)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:297)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.hazelcast.jet.impl.exception.TerminatedWithSnapshotException
	at com.hazelcast.jet.impl.execution.ExecutionContext.lambda$beginExecution$0(ExecutionContext.java:143)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
	... 8 more
13:51:45,677  INFO || - [MasterContext] hz._hzInstance_6_jet.async.thread-5 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Snapshot 0 for job '1d75-6ebe-5e2d-19f1', execution 1c1b-edaf-840f-df43 completed with status SUCCESS in 375ms, 89 bytes, 2 keys in 1 chunks, stored in '__jet.snapshot.1d75-6ebe-5e2d-19f1.0'
13:51:45,697 DEBUG || - [MasterContext] hz._hzInstance_6_jet.async.thread-7 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Execution of job '1d75-6ebe-5e2d-19f1', execution 1c1b-edaf-840f-df43 has failures: [MemberInfo{address=[127.0.0.1]:5701, uuid=6fc97810-399e-41e8-94a1-5c43b21ec846, liteMember=false, memberListJoinVersion=1}=com.hazelcast.jet.impl.exception.TerminatedWithSnapshotException]
13:51:45,823 DEBUG || - [MasterContext] hz._hzInstance_6_jet.async.thread-7 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Execution of job '1d75-6ebe-5e2d-19f1', execution 1c1b-edaf-840f-df43 terminated after a terminal snapshot
13:51:45,835 DEBUG || - [MasterContext] hz._hzInstance_6_jet.async.thread-7 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Sending CompleteExecutionOperation for job '1d75-6ebe-5e2d-19f1', execution 1c1b-edaf-840f-df43
13:51:45,915 DEBUG || - [CompleteExecutionOperation] hz._hzInstance_6_jet.async.thread-7 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Completing execution 1c1b-edaf-840f-df43 from caller [127.0.0.1]:5701, error=com.hazelcast.jet.impl.exception.JobTerminateRequestedException: RESTART_GRACEFUL
13:51:45,960 DEBUG || - [JobRepository] hz._hzInstance_6_jet.async.thread-5 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Cleared snapshot data map __jet.snapshot.1d75-6ebe-5e2d-19f1.1
13:51:45,961 DEBUG || - [MasterContext] hz._hzInstance_6_jet.async.thread-5 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Not beginning snapshot since terminal snapshot is already completed job '1d75-6ebe-5e2d-19f1', execution 1c1b-edaf-840f-df43
13:51:46,128 DEBUG || - [JobExecutionService] hz._hzInstance_6_jet.async.thread-7 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Completed execution of job '1d75-6ebe-5e2d-19f1', execution 1c1b-edaf-840f-df43
13:51:46,168  INFO || - [MasterContext] ForkJoinPool.commonPool-worker-57 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Execution of job '1d75-6ebe-5e2d-19f1', execution 1c1b-edaf-840f-df43 completed in 4,237 ms, reason=com.hazelcast.jet.impl.exception.JobTerminateRequestedException: RESTART_GRACEFUL
13:51:46,219  INFO || - [NodeExtension] Thread-0 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Destroying node NodeExtension.
13:51:46,254  INFO || - [Node] Thread-0 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] Hazelcast Shutdown is completed in 1004 ms.
13:51:46,254  INFO || - [LifecycleService] Thread-0 - [127.0.0.1]:5701 [jet] [0.8-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN
BuildInfo right after smokeTest(com.hazelcast.jet.impl.connector.StreamEventJournalPTest): BuildInfo{version='3.12-SNAPSHOT', build='20190212', buildNumber=20190212, revision=862df02, enterprise=false, serializationVersion=1, jet=JetBuildInfo{version='0.8-SNAPSHOT', build='20190213', revision='9c4082e'}}
Hiccups measured while running test 'smokeTest(com.hazelcast.jet.impl.connector.StreamEventJournalPTest):'
13:51:30, accumulated pauses: 2082 ms, max pause: 806 ms, pauses over 1000 ms: 0
13:51:35, accumulated pauses: 2503 ms, max pause: 876 ms, pauses over 1000 ms: 0
13:51:40, accumulated pauses: 2674 ms, max pause: 240 ms, pauses over 1000 ms: 0
13:51:45, accumulated pauses: 438 ms, max pause: 142 ms, pauses over 1000 ms: 0

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/test/TestSupport.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/test/TestSupport.java
@@ -50,10 +50,13 @@ import static com.hazelcast.jet.core.test.JetAssert.assertFalse;
 import static com.hazelcast.jet.core.test.JetAssert.assertTrue;
 import static com.hazelcast.jet.function.DistributedFunction.identity;
 import static com.hazelcast.jet.impl.util.ExceptionUtil.sneakyThrow;
+import static com.hazelcast.jet.impl.util.Util.subtractClamped;
+import static com.hazelcast.util.Preconditions.checkNotNegative;
 import static java.util.Collections.emptyList;
 import static java.util.Collections.singletonList;
 import static java.util.concurrent.TimeUnit.MICROSECONDS;
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
 import static java.util.stream.Collectors.toMap;
 
 /**
@@ -73,9 +76,10 @@ import static java.util.stream.Collectors.toMap;
  *
  *     <li>every time the inbox gets empty does snapshot or snapshot+restore
  *
- *     <li>{@link #disableCompleteCall() optionally} calls {@link Processor#complete()}
- *     until it returns {@code true} or calls it until {@link #disableRunUntilCompleted(long)
- *     specified timeout} elapses (for streaming sources)
+ *     <li>{@linkplain #disableCompleteCall() optionally} calls {@link
+ *     Processor#complete()} until it returns {@code true} or until the
+ *     {@linkplain #runUntilOutputMatches output matches} (for streaming
+ *     sources)
  *
  *     <li>does snapshot or snapshot+restore each time the {@code complete()}
  *     method returned {@code false} and made a progress
@@ -202,7 +206,8 @@ public final class TestSupport {
     private boolean callComplete = true;
     private JetInstance jetInstance;
     private long cooperativeTimeout = COOPERATIVE_TIME_LIMIT_MS_FAIL;
-    private long runUntilCompletedTimeout;
+    private long runUntilOutputMatchesTimeoutMillis = -1;
+    private long runUntilOutputMatchesExtraTimeMillis;
 
     private BiPredicate<? super List<?>, ? super List<?>> outputChecker = Objects::equals;
 
@@ -354,28 +359,29 @@ public final class TestSupport {
 
     /**
      * Normally, the {@code complete()} method is run repeatedly until it
-     * returns {@code true}. But in infinite source processors the method never
-     * returns {@code true}. To be able test such processors, this method
-     * allows you to disable the "run until completed" behavior and instead run
-     * the {@code complete()} for a specified time.
+     * returns {@code true}. This is suitable for processors processing the
+     * input or for batch sources. However, if you test a streaming source, the
+     * {@code complete()} method never returns {@code true}. To be able test
+     * such processors, this method allows you to change the behavior to run
+     * {@code complete()} until the output matches.
      * <p>
-     * If the timeout > 0, the {@code complete()} method is called repeatedly
-     * until the timeout elapses. After that, the output is compared using the
-     * {@link #outputChecker(BiPredicate) output checker}. The {@code
-     * complete()} method is also not allowed to return {@code true} in this
-     * case.
+     * The {@code extraTimeMillis} parameter specifies an extra time to call
+     * {@code complete()} after the output matches. It can be used to ensure
+     * that no more items are produced after the output matches.
      * <p>
-     * If the timeout is <= 0 (the default), {@code complete()} method is
-     * called until it returns {@code true}, after which the output is checked.
-     * <p>
-     * Has no effect if {@code complete()} call is {@link #disableCompleteCall()
-     * disabled}.
+     * Has no effect if calling {@code complete()} is {@linkplain
+     * #disableCompleteCall() disabled}.
      *
-     * @param timeoutMillis how long to wait until outputs match
+     * @param timeoutMillis maximum time to wait for the output to match
+     * @param extraTimeMillis for how long to call {@code complete()}
+     *                       after the output matches
      * @return {@code this} instance for fluent API
      */
-    public TestSupport disableRunUntilCompleted(long timeoutMillis) {
-        this.runUntilCompletedTimeout = timeoutMillis;
+    public TestSupport runUntilOutputMatches(long timeoutMillis, long extraTimeMillis) {
+        checkNotNegative(timeoutMillis, "timeoutMillis must be >= 0");
+        checkNotNegative(extraTimeMillis, "extraTimeMillis must be >= 0");
+        this.runUntilOutputMatchesTimeoutMillis = timeoutMillis;
+        this.runUntilOutputMatchesExtraTimeMillis = extraTimeMillis;
         return this;
     }
 
@@ -546,10 +552,10 @@ public final class TestSupport {
         // call the complete() method
         if (callComplete) {
             long completeStart = System.nanoTime();
+            long outputMatchedAt = Long.MAX_VALUE;
             boolean[] done = {false};
-            double elapsed;
             do {
-                checkTime("complete", isCooperative, () -> done[0] = processor[0].complete());
+                doCall("complete", isCooperative, () -> done[0] = processor[0].complete());
                 boolean madeProgress = done[0] || !outbox[0].queue(0).isEmpty();
                 assertTrue("complete() call without progress", !assertProgress || madeProgress);
                 outbox[0].drainQueuesAndReset(actualOutputs, logInputOutput);
@@ -562,19 +568,36 @@ public final class TestSupport {
                             doRestoreEvery, restoreCount);
                 }
                 idleCount = idle(idler, idleCount, madeProgress);
-                if (runUntilCompletedTimeout > 0) {
-                    elapsed = toMillis(System.nanoTime() - completeStart);
-                    if (elapsed > runUntilCompletedTimeout) {
+                long now = System.nanoTime();
+                if (runUntilOutputMatchesTimeoutMillis >= 0) {
+                    try {
+                        asssertOutput(doSnapshots, doRestoreEvery, inboxLimit, actualOutputs);
+                        outputMatchedAt = Math.min(outputMatchedAt, now);
+                    } catch (AssertionError e) {
+                        if (outputMatchedAt < Long.MAX_VALUE) {
+                            throw new AssertionError("the output already matched, but doesn't match now", e);
+                        }
+                        // ignore the failure otherwise and continue calling complete()
+                    }
+                    long elapsedSinceStart = NANOSECONDS.toMillis(now - completeStart);
+                    long elapsedSinceMatch = NANOSECONDS.toMillis(subtractClamped(now, outputMatchedAt));
+                    if (elapsedSinceStart > runUntilOutputMatchesTimeoutMillis
+                            || elapsedSinceMatch > runUntilOutputMatchesExtraTimeMillis) {
                         break;
                     }
                 }
             } while (!done[0]);
-            assertTrue("complete returned true", !done[0] || runUntilCompletedTimeout <= 0);
+            assertTrue("complete returned true in a run-until-output-matches mode",
+                    !done[0] || runUntilOutputMatchesTimeoutMillis <= 0);
         }
 
         processor[0].close();
 
         // assert the outbox
+        asssertOutput(doSnapshots, doRestoreEvery, inboxLimit, actualOutputs);
+    }
+
+    private void asssertOutput(boolean doSnapshots, int doRestoreEvery, int inboxLimit, List<List<Object>> actualOutputs) {
         for (int i = 0; i < expectedOutputs.size(); i++) {
             List<?> expectedOutput = expectedOutputs.get(i);
             List<?> actualOutput = actualOutputs.get(i);
@@ -627,14 +650,14 @@ public final class TestSupport {
     private String processInbox(TestInbox inbox, int inboxOrdinal, boolean isCooperative, Processor[] processor) {
         if (inbox.peek() instanceof Watermark) {
             Watermark wm = ((Watermark) inbox.peek());
-            checkTime("tryProcessWatermark", isCooperative, () -> {
+            doCall("tryProcessWatermark", isCooperative, () -> {
                 if (processor[0].tryProcessWatermark(wm)) {
                     inbox.remove();
                 }
             });
             return "tryProcessWatermark";
         } else {
-            checkTime("process", isCooperative, () -> processor[0].process(inboxOrdinal, inbox));
+            doCall("process", isCooperative, () -> processor[0].process(inboxOrdinal, inbox));
             return "process";
         }
     }
@@ -672,7 +695,7 @@ public final class TestSupport {
         boolean[] done = {false};
         boolean isCooperative = processor[0].isCooperative();
         do {
-            checkTime("saveSnapshot", isCooperative, () -> done[0] = processor[0].saveToSnapshot());
+            doCall("saveSnapshot", isCooperative, () -> done[0] = processor[0].saveToSnapshot());
             assertTrue("saveToSnapshot() call without progress",
                     !assertProgress || done[0] || !outbox[0].snapshotQueue().isEmpty()
                             || !outbox[0].queue(0).isEmpty());
@@ -694,7 +717,7 @@ public final class TestSupport {
 
         int lastInboxSize = snapshotInbox.queue().size();
         while (!snapshotInbox.isEmpty()) {
-            checkTime("restoreSnapshot", isCooperative,
+            doCall("restoreSnapshot", isCooperative,
                     () -> processor[0].restoreFromSnapshot(snapshotInbox));
             assertTrue("restoreFromSnapshot() call without progress",
                     !assertProgress
@@ -704,7 +727,7 @@ public final class TestSupport {
             lastInboxSize = snapshotInbox.queue().size();
         }
         do {
-            checkTime("finishSnapshotRestore", isCooperative,
+            doCall("finishSnapshotRestore", isCooperative,
                     () -> done[0] = processor[0].finishSnapshotRestore());
             assertTrue("finishSnapshotRestore() call without progress",
                     !assertProgress || done[0] || !outbox[0].queue(0).isEmpty());
@@ -712,7 +735,7 @@ public final class TestSupport {
         } while (!done[0]);
     }
 
-    private void checkTime(String methodName, boolean isCooperative, Runnable r) {
+    private void doCall(String methodName, boolean isCooperative, Runnable r) {
         long start = System.nanoTime();
         r.run();
         long elapsed = System.nanoTime() - start;

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MigrationWatcher.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MigrationWatcher.java
@@ -44,7 +44,7 @@ public class MigrationWatcher {
      */
     public BooleanSupplier createWatcher() {
         int startChangeCount = changeCount.get();
-        return () -> changeCount.get() - startChangeCount > 0; // this is overflow-safe
+        return () -> changeCount.get() != startChangeCount;
     }
 
     public void deregister() {

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/Util.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/Util.java
@@ -269,6 +269,14 @@ public final class Util {
                 : sum;
     }
 
+    /**
+     * Calculates {@code a - b}, returns {@code Long.MAX_VALUE} if the result
+     * would overflow
+     *
+     * @param a the amount
+     * @param b the value to subtract
+     * @return {@code a - b}, clamped
+     */
     public static long subtractClamped(long a, long b) {
         long diff = a - b;
         return diffHadOverflow(a, b, diff)

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/StreamEventJournalPTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/StreamEventJournalPTest.java
@@ -102,7 +102,7 @@ public class StreamEventJournalPTest extends JetTestSupport {
 
         TestSupport.verifyProcessor(supplier)
                    .disableProgressAssertion() // no progress assertion because of async calls
-                   .disableRunUntilCompleted(1000) // processor would never complete otherwise
+                   .runUntilOutputMatches(60_000, 100)
                    .outputChecker(SAME_ITEMS_ANY_ORDER) // ordering is only per partition
                    .expectOutput(Arrays.asList(0, 1, 2, 3));
     }

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/StreamEventJournalP_WmCoalescingTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/StreamEventJournalP_WmCoalescingTest.java
@@ -89,7 +89,7 @@ public class StreamEventJournalP_WmCoalescingTest extends JetTestSupport {
 
         TestSupport.verifyProcessor(createSupplier(asList(0, 1), 2000))
                    .disableProgressAssertion()
-                   .disableRunUntilCompleted(1000)
+                   .runUntilOutputMatches(60_000, 100)
                    .disableSnapshots()
                    .expectOutput(asList(wm(10), 10, 10));
     }
@@ -114,7 +114,7 @@ public class StreamEventJournalP_WmCoalescingTest extends JetTestSupport {
 
         TestSupport.verifyProcessor(createSupplier(asList(0, 1), 4000))
                    .disableProgressAssertion()
-                   .disableRunUntilCompleted(4000)
+                   .runUntilOutputMatches(60_000, 100)
                    .disableSnapshots()
                    .outputChecker((e, a) -> new HashSet<>(e).equals(new HashSet<>(a)))
                    .expectOutput(asList(11, wm(11)));
@@ -126,7 +126,7 @@ public class StreamEventJournalP_WmCoalescingTest extends JetTestSupport {
     public void when_allPartitionsIdle_then_idleMessageOutput() {
         TestSupport.verifyProcessor(createSupplier(asList(0, 1), 500))
                    .disableProgressAssertion()
-                   .disableRunUntilCompleted(1500)
+                   .runUntilOutputMatches(60_000, 100)
                    .disableSnapshots()
                    .expectOutput(singletonList(IDLE_MESSAGE));
     }
@@ -170,7 +170,7 @@ public class StreamEventJournalP_WmCoalescingTest extends JetTestSupport {
 
         TestSupport.verifyProcessor(createSupplier(singletonList(1), 2000))
                    .disableProgressAssertion()
-                   .disableRunUntilCompleted(4000)
+                   .runUntilOutputMatches(60_000, 100)
                    .disableSnapshots()
                    .expectOutput(asList(wm(13), 13, IDLE_MESSAGE));
     }

